coordinators/simple/coordinator.go

```diff
@@ -445,79 +459,73 @@ func (c *Simple) PerformFlowIntegrityCheck(fid iface.FlowID) (iface.FlowDataInte
-		slog.Error("Failed to request integrity check from source", err)
-		return res, err
-	}
-	if err := dst.Endpoint.RequestDataIntegrityCheck(fid, flowDet.Options.SrcConnectorOptions); err != nil { //TODO (AK, 6/2024): should we have proper options here? (maybe even data validation-specific?)
-		slog.Error("Failed to request integrity check from destination", err)
+	if err := c.maybeCreateReadPlan(src.Endpoint, flowDet); err != nil {
 		return res, err
 	}
 
-	// Wait for both results
-	select {
-	case resSource = <-flowDet.integrityCheckDoneChannels[0]:
-		slog.Debug("Got integrity check result from source: " + fmt.Sprintf("%v", resSource))
-	case <-c.ctx.Done():
-		slog.Debug("Context cancelled. Flow ID: " + fmt.Sprintf("%v", fid))
-	}
-	select {
-	case resDestination = <-flowDet.integrityCheckDoneChannels[1]:
-		slog.Debug("Got integrity check result from destination: " + fmt.Sprintf("%v", resDestination))
-	case <-c.ctx.Done():
-		slog.Debug("Context cancelled. Flow ID: " + fmt.Sprintf("%v", fid))
+	numWorkers := c.numParallelIntegrityCheckTasks
+	if numWorkers < 1 {
+		numWorkers = len(flowDet.ReadPlan.Tasks)
 	}
+	taskCh := make(chan iface.ReadPlanTask)
+	mismatchErr := errors.New("match failed")
 
-	if (resSource == iface.ConnectorDataIntegrityCheckResult{}) || (resDestination == iface.ConnectorDataIntegrityCheckResult{}) {
-		slog.Debug("Integrity check results are empty")
-		return res, fmt.Errorf("integrity check results are empty")
-	}
+	go func() {
+		defer close(taskCh)
+		for _, task := range flowDet.ReadPlan.Tasks {
+			select {
+			case taskCh <- task:
+			case <-ctx.Done():
+				slog.Info(fmt.Sprintf("Integrity task canceled %v", task))
+				return
+			}
+		}
+	}()
 
-	if (!resSource.Success) || (!resDestination.Success) {
-		slog.Debug("Integrity check failure on either end")
-		return res, fmt.Errorf("integrity check failure on either end")
+	var eg errgroup.Group
+	for i := 0; i < numWorkers; i++ {
+		eg.Go(func() error {
+			for task := range taskCh {
+				srcRes, err := src.Endpoint.IntegrityCheck(ctx, fid, task)
+				if err != nil {
+					if errors.Is(err, context.Canceled) {
+						slog.Info(fmt.Sprintf("Source integrity task canceled %v", task))
+						return nil
+					}
+					slog.Error(fmt.Sprintf("Source integrity check failed: %v", err))
+					return err
+				}
+				dstRes, err := dst.Endpoint.IntegrityCheck(ctx, fid, task)
+				if err != nil {
+					if errors.Is(err, context.Canceled) {
+						slog.Info(fmt.Sprintf("Destination integrity task canceled %v", task))
+						return nil
+					}
+					slog.Error(fmt.Sprintf("Destination integrity check failed: %v", err))
+					return err
+				}
+				matches := bytes.Equal(srcRes.Digest, dstRes.Digest) && srcRes.Count == dstRes.Count
+				if !matches {
+					slog.Info(fmt.Sprintf("Mismatch %v, src: %v, dest: %v", task, srcRes, dstRes))
+					return mismatchErr
+				}
+				slog.Info(fmt.Sprintf("Matched %v", task))
```

Comment:
This probably should be DEBUG

Reply:
this is essentially the poor man's progress report